staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go

```diff
@@ -369,3 +369,76 @@ func TestFilterPartialMatchesSometimesMissingAnnotations(t *testing.T) {
 		t,
 	)
 }
+
+func TestNegativeIndex(t *testing.T) {
+	var input = []byte(
+		`{
+			"apiVersion": "v1",
+			"kind": "Pod",
+			"spec": {
+				"containers": [
+					{
+						"image": "radial/busyboxplus:curl",
+						"name": "fake0"
+					},
+					{
+						"image": "radial/busyboxplus:curl",
+						"name": "fake1"
+					},
+					{
+						"image": "radial/busyboxplus:curl",
+						"name": "fake2"
+					},
+					{
+						"image": "radial/busyboxplus:curl",
+						"name": "fake3"
+					}]}}`)
+
+	var data interface{}
+	err := json.Unmarshal(input, &data)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testJSONPath(
```

Comment:
thanks for these... some additional cases are needed around range behavior (which the current fix changes):

```
	testJSONPath(
		[]jsonpathTest{
			{
				"test containers[0], it equals containers[0]",
				`{.spec.containers[0].name}`,
				data,
				"fake0",
				false,
			},
			{
				"test containers[0:0], it equals the empty set",
				`{.spec.containers[0:0].name}`,
				data,
				"",
				false,
			},
			{
				"test containers[0:-1], it equals containers[0:3]",
				`{.spec.containers[0:-1].name}`,
				data,
				"fake0 fake1 fake2",
				false,
			},
			{
				"test containers[-1:0], it equals containers[3:0]",
				`{.spec.containers[-1:0].name}`,
				data,
				"fake0 fake1 fake2",
				false,
			},
			{
				"test containers[-1], it equals containers[3]",
				`{.spec.containers[-1].name}`,
				data,
				"fake3",
				false,
			},
			{
				"test containers[-1:], it equals containers[3:]",
				`{.spec.containers[-1:].name}`,
				data,
				"fake3",
				false,
			},
			{
				"test containers[-2], it equals containers[2]",
				`{.spec.containers[-2].name}`,
				data,
				"fake2",
				false,
			},
			{
				"test containers[-2:], it equals containers[2:]",
				`{.spec.containers[-2:].name}`,
				data,
				"fake2 fake3",
				false,
			},
			{
				"test containers[-3], it equals containers[1]",
				`{.spec.containers[-3].name}`,
				data,
				"fake1",
				false,
			},
			{
				"test containers[-4], it equals containers[0]",
				`{.spec.containers[-4].name}`,
				data,
				"fake0",
				false,
			},
			{
				"test containers[-4:], it equals containers[0:]",
				`{.spec.containers[-4:].name}`,
				data,
				"fake0 fake1 fake2 fake3",
				false,
			},
			{
				"test containers[-5], expect a error cause it out of bounds",
				`{.spec.containers[-5].name}`,
				data,
				"",
				true, // expect error
			},
		},
		false,
		t,
	)
```

Reply:
also, catching panics in the test makes failures easier to see:
```
func testJSONPath(tests []jsonpathTest, allowMissingKeys bool, t *testing.T) {
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			defer func() {
				if err := recover(); err != nil {
					t.Error(err)
				}
			}()
			j := New(test.name)
			j.AllowMissingKeys(allowMissingKeys)
			err := j.Parse(test.template)
			if err != nil {
				t.Errorf("in %s, parse %s error %v", test.name, test.template, err)
			}
			buf := new(bytes.Buffer)
			err = j.Execute(buf, test.input)
			if test.expectError {
				if test.expectError && err == nil {
					t.Errorf("in %s, expected execute error", test.name)
				}
				return
			} else if err != nil {
				t.Errorf("in %s, execute error %v", test.name, err)
			}
			out := buf.String()
			if out != test.expect {
				t.Errorf(`in %s, expect to get "%s", got "%s"`, test.name, test.expect, out)
			}
		})
	}
}
```